### ISSUE TITLE
ConfigSourceFactory not honoring profile if using profile specific files

### ISSUE THREAD

**Issue by manofthepeace** [smallrye/smallrye-config#752]
This is a duplicate of https://github.com/quarkusio/quarkus/issues/25429

I would think this is more of a smallrye config issue than a quarkus one. I can close one or the other if needed.

I have the reproducer here; https://github.com/manofthepeace/quarkus-configsourcefactory

Thank you for your time.

**Comment by radcortez** [comment]
@manofthepeace sorry for the late response.

This happens because the Profile sources are also implemented with Factories and they are only loaded after the main sources are evaluated. One of the main reasons is that we need to determine the profile to use (which may be set in any of the main sources) and only after that, we are able to load them.

To support this use case, we need to split the loading process a bit more to prioritize and add to the source chain the profile sources and afterward load the sources coming from user Factories.

**Comment by manofthepeace** [comment]
Thank you for your reply. Great explanation as always and no worries for the delay, I know you guys are extremely busy.

I feel like in my use case, I am a bit going overkill with the factory, but I haven't found another way. Maybe you can tell me if what I am doing is ok in the first place. It does work great I must say, but I feel the thing is a bit involved for what I actually need.

Basically, what I need is to be able to overwrite a config value, using the result of an other config. Generally in spring, I would use SpEL for that; in which I can check values compare,concat etc.

So I have `@Scheduled` methods, with configured TemporalAmount in my config files, and also I have other config as feature flag. if the featureX is disabled. I use the factory, to overwrite the property via a new PropertiesConfigSource to "off" with a higher ordinal, so the scheduler is not enabled for these tasks.

basically like this;
prop;
```properties
my.feature=false
my.task.interval=10m
```

ConfigSourceFactory . I check if context.getValue("my.feature") is false then if true
```java
Map<String, String> propMap = Map.of("my.task.interval", "off");
return Collections.singletonList(new PropertiesConfigSource(propMap, "feature-flags", 500));
```

**Comment by radcortez** [comment]
Yes, that is certainly one way to do it and it works.

An alternative would be to use Interceptors:
https://smallrye.io/smallrye-config/2.10.0/extensions/interceptors/

Interceptors are applied after a property is evaluated. We use this to relocate properties or fallback to other ones:

https://github.com/smallrye/smallrye-config/blob/main/implementation/src/main/java/io/smallrye/config/RelocateConfigSourceInterceptor.java
https://github.com/smallrye/smallrye-config/blob/main/implementation/src/main/java/io/smallrye/config/FallbackConfigSourceInterceptor.java

So, in this case, when someone is requesting a property that needs to be verified with a feature flag, you could check the flag and then override the value returned by the original property on the fly.

Anyway, I should be able to find a way to support profile-aware sources at the user Factory level.

**Comment by manofthepeace** [comment]
Thanks for the reply. I have just tested the interceptor and it seems to work great as well, so I can put back the configs back in their profile-aware file. Just one thing I got that i'd like to note, in the ConfigSourceInterceptor, if I use this;

```java
    @ConfigProperty(name = "my.feature")
    boolean featureFlag;
```

it works fine; but if within the `getValues()` method I do the following;
`ConfigProvider.getConfig().getConfigValue("my.feature").getValue();`

Sometimes it works, sometimes it gives me the same kind of exception when I am trying to use a forkjoinpool (see https://github.com/quarkusio/quarkus/issues/24896)
```
Caused by: java.util.ServiceConfigurationError: io.smallrye.config.SmallRyeConfigFactory: io.quarkus.runtime.configuration.QuarkusConfigFactory not a subtype
	at java.base/java.util.ServiceLoader.fail(ServiceLoader.java:593)
	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNextService(ServiceLoader.java:1244)
	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNext(ServiceLoader.java:1273)
	at java.base/java.util.ServiceLoader$2.hasNext(ServiceLoader.java:1309)
	at java.base/java.util.ServiceLoader$3.hasNext(ServiceLoader.java:1393)
	at io.smallrye.config.SmallRyeConfigProviderResolver.getFactoryFor(SmallRyeConfigProviderResolver.java:100)
	at io.smallrye.config.SmallRyeConfigProviderResolver.getConfig(SmallRyeConfigProviderResolver.java:76)
	at io.smallrye.config.SmallRyeConfigProviderResolver.getConfig(SmallRyeConfigProviderResolver.java:64)
	at org.eclipse.microprofile.config.ConfigProvider.getConfig(ConfigProvider.java:85)
```

to reproduce everytime I just have to go to this url; http://localhost:8080/q/dev/io.quarkus.quarkus-scheduler/schedules

So I guess I will stay with the `@ConfigProperty` as it seems to not cause the exception.

**Comment by manofthepeace** [comment]
@radcortez in fact I have spoken too fast. even with `@ConfigProperty` I cannot get reliable result as the interceptor is called a lot of times for my value, and sometimes the value from the configProperty is null for some reason.

So would you know how from the interceptor, I can reliably get another config value?

my interceptor looks like this at the moment;

```java
    @ConfigProperty(name = "my.feature")
    String featureFlag;

    @Override
    public ConfigValue getValue(ConfigSourceInterceptorContext context, String name) {
        if (name.equals("my.task.interval")) {
            // String featureFlag = ConfigProvider.getConfig().getConfigValue("my.feature").getValue();
            if (!Boolean.parseBoolean(featureFlag)) {
                ConfigValue value = context.proceed(name);
                return value.withValue("off");
            }
        }
        return context.proceed(name);
    }
```

**Comment by radcortez** [comment]
My guess, considering the previous Exception and what you described in https://github.com/quarkusio/quarkus/issues/24896, is that you are calling the Config code in a Classloader that has none associated. In that case, the Config instance will automatically create one for you (with default settings), so you may be seeing this strange behavior.

Can you please update your reproducer with the interceptor and the exact code you are running? Thanks!

**Comment by manofthepeace** [comment]
I sure can. here you go https://github.com/manofthepeace/srconfig-interceptor-issue

**Test 1 (with ConfigProperty)**
1. run the app via run mvn quarkus:dev
2. see
`2022-05-24 10:42:13,467 INFO  [org.acm.ConfigInterceptor] (Quarkus Main Thread) feature.enabled is: null`

This is using the `@ConfigProperty`. In my real app I see it evaluated for some reason like 5 times, and sometimes it's ok, sometimes it's null as above.

**Test 2 (With ConfigProvider)**
1. Comment out the configProperty+its var at the top
2. uncomment `// String featureFlag = ConfigProvider.getConfig().getConfigValue("my.feature.enabled").getValue();`
3. run mvn quarkus:dev
4. go to `http://localhost:8080/q/dev/io.quarkus.quarkus-scheduler/schedules`
5. Exception happens

I am a bit troubled with the exception sine it worries me things can happen at runtime/prod, since I do not know what exactly triggers it. I understand a little the classloading issue, but in the sample provided, it's all internal to quarkus, not under my control I believe.

**Comment by radcortez** [comment]
I had a look into the code, and the issue is that the interceptor is a low-level API. It doesn't support CDI (and injection), because we don't want to require CDI to use it.

Also, using `ConfigProvider.getConfig()` in the interceptor code is not recommended. It can lead to infinite loops or stackoverflow exceptions, because you are calling the interceptor again. In this. case, the error is an issue with the CL related to the dev console, which I'll have to look at and fix.

The recommendation is to call `context.proceed()` with the new key. So you can write the code like this and it works:

```java
@Override
    public ConfigValue getValue(ConfigSourceInterceptorContext context, String name) {
        if (name.equals("my.feature.delay")) {
            ConfigValue featureFlag = context.proceed("my.feature.enabled");
            if (featureFlag != null && featureFlag.getValue() != null) {
                LOG.infof("feature.enabled is: %s", featureFlag.getValue());
                if (!Boolean.parseBoolean(featureFlag.getValue())) {
                    return context.proceed(name).withValue("off");
                }
            }
        }
        return context.proceed(name);
    }
```

In the meanwhile, I've already created a PR to allow the profiles values to be available in the Factory context.
